Formula/netdata.rb

```diff
@@ -29,6 +31,26 @@ class Netdata < Formula
     sha256 "d2704089f85fdb6f2cd7e77be21170ced4b4375c03ef1ad4cf1075bd414a63eb"
   end
 
+  # Requires pinned, vendored libwebsockets to support ACLK
+  # See https://github.com/netdata/netdata/issues/9061
+  resource "libwebsockets" do
+    # https://github.com/netdata/netdata/blob/v1.24.0/packaging/libwebsockets.version
+    url "https://github.com/warmcat/libwebsockets/archive/v3.2.2.tar.gz"
+
+    # https://github.com/netdata/netdata/blob/v1.24.0/packaging/libwebsockets.checksums
+    sha256 "166d6e17cab64bfc10c2a71799c298284540a1fa63f6ea3de5caccb34502243c"
+  end
+
+  # Requires patched libmosquitto to support ACLK
+  # See https://github.com/netdata/netdata/issues/9061
+  resource "mosquitto" do
+    # https://github.com/netdata/netdata/blob/master/packaging/mosquitto.version
+    url "https://github.com/netdata/mosquitto/archive/v.1.6.8_Netdata-4.tar.gz"
+
+    # https://github.com/netdata/netdata/blob/master/packaging/mosquitto.checksums
+    sha256 "6af837c388b1bcd459220936c422b70a987c60f4d5f88b09eada43d713529284"
+  end
```

Comment:
These should really be made to work with the homebrew formula. We don't want to vendor anything we have a formula for.

Reply:
libmosquito has to be vendored at the moment, we need a custom patch on top of upstream for it to work. We're working on getting that upstream, but I don't know of any ETA on that.

LWS, OTOH, could theoretically be handled with a system copy (provided the system copy is built with SOCKS5 support), but our build system currently does not support that approach.

We have an internal issue tracking support for this here: https://github.com/netdata/netdata/issues/8961

Reply:
In that case we should keep cloud integration disabled until it can meet homebrew requirements.

Reply:
> We don't want to vendor anything we have a formula for.

I understand this argument I think it's a bit arbitrary since for Python and Ruby formulas this is exactly what is done (in contrary with how the "native" packaging/dependency management work).

I would say that in this case it should be allowed since this statically links these dependencies into the binary and Netdata is very actively maintained will likely resolve these issues in the future and then remove the vendoring.

Also while checking the guidelines for acceptable formals etc I couldn't find any mentioning that this is explicitly disallowed.

Reply:
> I understand this argument I think it's a bit arbitrary since for Python and Ruby formulas this is exactly what is done (in contrary with how the "native" packaging/dependency management work).

I'm not sure what you mean but nothing here is vendored: https://github.com/Homebrew/homebrew-core/blob/master/Formula/ruby.rb

> I would say that in this case it should be allowed since this statically links these dependencies into the binary and Netdata is very actively maintained will likely resolve these issues in the future and then remove the vendoring.

Looking at the open issue that has been parked for months doesn't fill me with confidence that this will be resolved any time soon.

> Also while checking the guidelines for acceptable formals etc I couldn't find any mentioning that this is explicitly disallowed.

You're right, I'll make a pull request to improve that.

Reply:
> I'm not sure what you mean but nothing here is vendored: https://github.com/Homebrew/homebrew-core/blob/master/Formula/ruby.rb

Like this https://github.com/Homebrew/homebrew-core/blob/master/Formula/awscli.rb#L28-L30

> Looking at the open issue that has been parked for months doesn't fill me with confidence that this will be resolved any time soon.

I specifically avoided the word soon. :wink: 

> You're right, I'll make a pull request to improve that.

:+1:

Reply:
> Like this https://github.com/Homebrew/homebrew-core/blob/master/Formula/awscli.rb#L28-L30

We vendor python dependencies cause pip...is difficult to work with as a package manager. If python libraries worked the same way system libraries do we could remove a lot of duplication.

Reply:
I'm just going to say this once. I _really_ don't understand why in general (_this isn't aimed at Homebrew core at all!_) we get a bit zealous with packaging and vendoring. A package manager is a tool, we use it as such to "package software". However that is accomplished is really not the point here there are _often_ **very good** reasons to vendor a dependency or pull in sources from other sources (_depending on licensing/needs_).

As my colleague @Ferroin points out we _cannot_ current ship a Netdata Agent built with cloud support with _any_ system copy of the libwebsockets until we get the patches we had to make upstream (_currently no EAT on this_).

If we can't get this PR into Homebrew core, then we will remove it and offer Netdata as a Tap instead at https://github.com/netdata.

Reply:
I understand that sometimes there are patches that need to be done and homebrew will happily accept patches that make software work better for everyone. But if someone finds a CVE in libwebsockets we would really rather patch that in one place for all formula, rather than going through all of them and reminding authors that they need to update their vendored dependencies. If you can patch the current libwebsockets formula with some PR you made upstream that isn't released yet, that would be the best way forward IYAM.

As for removing the formula, we won't do that. This is open source software and homebrew is free to distribute it.
https://docs.brew.sh/How-to-Create-and-Maintain-a-Tap#official-vendor-taps

Reply:
@SMillerDev Thanks for your understanding and empathy. We _are_ trying out best to get our patches upstream. In the meantime we are a bit stuck.. I _get_ the whole "security" argument, however the argument can be made that it is just as much as "our" responsibility to patch any source, vendor'd libraries, etc as a company as it is for maintainers of open-source packages. In any case, I'll let you guys decide what you want to do here, We _cannot_ easily get the required pathces upstream in any reasonable timeframe I'm afraid, so bare that in mind.

Reply:
I think it's both our responsibilities indeed. You have the responsibility to your customers and homebrew has the responsibility to its users.

It's it's okay if you can't get those patches upstreamed, but in that case I'd like to hold off on adding the cloud feature to the homebrew formula until you can.

Reply:
Documentation is now available mentioning our hesitance to include vendored dependencies: https://docs.brew.sh/Acceptable-Formulae#stuff-that-requires-vendored-versions-of-homebrew-formulae

Should I close this PR so you can make a new one when it meets the requirements for properly using system dependencies?

Reply:
Thanks for your feedback and time taken to document Homebrew's position and your understanding. Yes I _think_ we can close this PR for the time being as I don't believe we can realistically get these patches upstream right now. Thanks!